### PR TITLE
Final cleanup

### PR DESCRIPTION
**Pull request by menshikh-iv**

### CODE REVIEW

File: generate_table.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    print("| name | description | papers | preprocessing | parameters |")
+    print("|------|-------------|------------|--------|---------------|")
+    for name, other in models:
+        if name.startswith("__testing_"):
+            continue
+        print("| {name} | {description} | {papers} | {preprocessing} | {parameters} |".format(
+            name=name, description=other["description"], parameters=other.get("parameters", "-"),
+            papers=other.get("papers", "-"), preprocessing=other.get("preprocessing", "-")
+        ))
+    print("(generated by generate_table.py based on lists.json)")
```
Comment (piskvorky) [inline generate_table.py]: Shouldn't this be `based on %{fn}` instead?

File: generate_table.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python
+"""Script for generation table with data/models from list.json in .md format"""
```
Comment (piskvorky) [inline generate_table.py]: Other places mention `lists.json`.

File: generate_table.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        print("| {name} | {description} | {papers} | {preprocessing} | {parameters} |".format(
+            name=name, description=other["description"], parameters=other.get("parameters", "-"),
+            papers=other.get("papers", "-"), preprocessing=other.get("preprocessing", "-")
+        ))
+    print("(generated by generate_table.py based on lists.json)")
+
+
+def main():
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument("input_file", help="Path to 'list.json' from repo")
```
Comment (piskvorky) [inline generate_table.py]: Other places mention `lists.json`.

**Comment by menshikh-iv** [comment]
Other fixes:
- Move developer info to wiki (no need to show it in README)
- Add info about "where current data stored" to README
- Add `api.info` output example
- Add `smart_open` for datasets

File: generate_table.py (hunk 2 above)
Comment (menshikh-iv) [reply generate_table.py]: This is the right option

File: generate_table.py (hunk 3 above)
Comment (menshikh-iv) [reply generate_table.py]: This is the right option, need to fix other places.

File: generate_table.py (hunk 1 above)
Comment (menshikh-iv) [reply generate_table.py]: I strongly prefer new `.format`, instead of C-style formatting.

Comment (menshikh-iv) [reply generate_table.py]: About fn - maybe, but file never changed

Comment (piskvorky) [reply generate_table.py]: OK, I meant outputting the actual parameter rather hardwiring some "magic" file name.

Btw magic vars are best defined at global scope (`DATASETS_SOURCE = "lists.json"`). It makes it easier to comment / explain / maintain / modify in the future.

**Comment by piskvorky** [comment]
@menshikh-iv please give me a ping when this is ready from your side; I'll polish the English a bit before the release. Thanks!

File: generate_table.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    print("| name | description | papers | preprocessing | parameters |")
+    print("|------|-------------|------------|--------|---------------|")
+    for name, other in models:
+        if name.startswith("__testing_"):
+            continue
+        print("| {name} | {description} | {papers} | {preprocessing} | {parameters} |".format(
+            name=name, description=other["description"], parameters=other.get("parameters", "-"),
+            papers=other.get("papers", "-"), preprocessing=other.get("preprocessing", "-")
+        ))
+    print("\n(generated by {} based on {})".format(__file__, os.path.basename(fn)))
```
Comment (piskvorky) [inline generate_table.py]: No need to hide the path -- very useful for debugging and clarity (multiple Python installations by mistake etc).

File: README.md [hunk 5]
```diff
@@ -57,6 +57,16 @@ To load model/dataset you can use 2 types of API:
  
   """
   ```
+  
+  Example: how to **only** download data (without loading to memory)
+  ```
+  import gensim.downloader as api
+  
+  print(api.load("20-newsgroups", return_path=True))  # output: /home/user/gensim-data/20-newsgroups/20-newsgroups.gz
+  print(api.load("glove-twitter-25"))  # output: /home/ivan/gensim-data/glove-twitter-25/glove-twitter-25.gz
```
Comment (piskvorky) [inline README.md]: No `return_path`?

File: generate_table.py [hunk 6]
```diff
@@ -37,7 +37,7 @@ def generate_table(fn):
             name=name, description=other["description"], parameters=other.get("parameters", "-"),
             papers=other.get("papers", "-"), preprocessing=other.get("preprocessing", "-")
         ))
-    print("\n(generated by {} based on {})".format(__file__, os.path.basename(fn)))
+    print("\n(generated by {} based on {})".format(os.path.abspath(__file__), os.path.abspath(fn)))
```
Comment (piskvorky) [inline generate_table.py]: Sorry to nitpick, but mirroring the parameter exactly (without any normpath / abspath) is better.

Every extra step between what the computer is doing and what the user is seeing in logs, is an unnecessary obfuscation, a source of misunderstandings and errors.

Comment (menshikh-iv) [reply generate_table.py]: Everything is fine, I did not understand you right away, now this fixed.

File: README.md [hunk 7]
```diff
@@ -1,23 +1,110 @@
...
+ 
+  """
+  ```
+  
+  Example: how to **only** download data (without loading to memory)
+  ```
+  import gensim.downloader as api
+  
+  print(api.load("20-newsgroups", return_path=True))  # output: /home/user/gensim-data/20-newsgroups/20-newsgroups.gz
+  print(api.load("glove-twitter-25", return_path=True))  # output: /home/user/gensim-data/glove-twitter-25/glove-twitter-25.gz
```
Comment (piskvorky) [inline README.md]: Missing file extension -- if this is a text file, should be `.txt.gz` (here and elsewhere).

By convention, files without an extension are binary files (which is not the case here).

File: _test_all.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+failed_datasets = []
+for idx, dataset_name in enumerate(info["corpora"]):
+    logger.info("{} ({} of {})".format(dataset_name, idx + 1, len(info["corpora"])))
+    try:
+        res = sum(1 for _ in api.load(dataset_name))
+        if res == 0:
+            raise Exception("empty dataset")
+
+    except Exception as e:
+        logger.critical(e)
```
Comment (piskvorky) [inline _test_all.py]: `logger.exception` better? (to also record the stack trace)

File: README.md (hunk 7 above)
Comment (menshikh-iv) [reply README.md]: It's impossible right now, we used common substitution like `<data-name>.gz` to retrieve a link.

Comment (piskvorky) [reply README.md]: OK.

**Pull request merged**